### PR TITLE
update libvirt-go and libvirt version to latest ( 6.5.0 ) and new docker build files

### PR DESCRIPTION
**Pull request by yb01**
## What is in this PR
Update runtime code to latest libvirt-go  and update libvirtd to 6.5.0

## Why this PR is needed
this is needed for VM vertical scaling feature, for hot-plugin CPU and Memory.

## Special note to CR:
This is being tested and once tested, the build will enabled using the new runtime image

## Testing:
Being able to build the runtime image with qemu 5.0.0 and libvirt 6.5.0. VM can be created with this image:

> 
> virsh # version
> Compiled against library: libvirt 6.5.0
> Using library: libvirt 6.5.0
> Using API: QEMU 6.5.0
> Running hypervisor: QEMU 5.0.0
> 
> virsh # list
>  Id   Name                       State
> ------------------------------------------
>  1    virtlet-b28bebf5-38ce-vm   running
> 
> virsh # dominfo 1
> Id:             1
> Name:           virtlet-b28bebf5-38ce-vm
> UUID:           b28bebf5-38ce-5f2c-686b-2bd4da91f6f9
> OS Type:        hvm
> State:          running
> CPU(s):         3
> CPU time:       11.3s
> Max memory:     200704 KiB
> Used memory:    200704 KiB
> Persistent:     yes
> Autostart:      disable
> Managed save:   no
> Security model: none
> Security DOI:   0
> 
> virsh # dumpxml 1
> <domain type='qemu' id='1' xmlns:qemu='http://libvirt.org/schemas/domain/qemu/1.0'>
>   <name>virtlet-b28bebf5-38ce-vm</name>
>   <uuid>b28bebf5-38ce-5f2c-686b-2bd4da91f6f9</uuid>
>   <maxMemory slots='16' unit='KiB'>409600</maxMemory>
>   <memory unit='KiB'>200704</memory>
>   <currentMemory unit='KiB'>200704</currentMemory>
>   <vcpu placement='static' current='3'>6</vcpu>
>   <cputune>
>     <shares>3072</shares>
>     <period>100000</period>
>     <quota>100000</quota>
>   </cputune>
>   <resource>
>     <partition>/machine</partition>
>   </resource>
>   <os>
>     <type arch='x86_64' machine='pc-i440fx-2.11'>hvm</type>
>     <boot dev='hd'/>
>   </os>
>   <features>
>     <acpi/>
>   </features>
>   <cpu mode='custom' match='exact' check='full'>
>     <model fallback='forbid'>EPYC</model>
>     <vendor>AMD</vendor>
>     <feature policy='require' name='acpi'/>
>     <feature policy='require' name='ss'/>
>     <feature policy='require' name='hypervisor'/>
>     <feature policy='require' name='erms'/>
>     <feature policy='require' name='mpx'/>
>     <feature policy='require' name='pcommit'/>
>     <feature policy='require' name='clwb'/>
>     <feature policy='require' name='pku'/>
>     <feature policy='require' name='la57'/>
>     <feature policy='require' name='3dnowext'/>
>     <feature policy='require' name='3dnow'/>
>     <feature policy='disable' name='vme'/>
>     <feature policy='disable' name='fma'/>
>     <feature policy='disable' name='avx'/>
>     <feature policy='disable' name='f16c'/>
>     <feature policy='disable' name='rdrand'/>
>     <feature policy='disable' name='avx2'/>
>     <feature policy='disable' name='rdseed'/>
>     <feature policy='disable' name='sha-ni'/>
>     <feature policy='disable' name='xsavec'/>
>     <feature policy='disable' name='fxsr_opt'/>
>     <feature policy='disable' name='misalignsse'/>
>     <feature policy='disable' name='3dnowprefetch'/>
>     <feature policy='disable' name='osvw'/>
>     <numa>
>       <cell id='0' cpus='0-5' memory='200704' unit='KiB'/>
>     </numa>
>   </cpu>
>   <clock offset='utc'/>
>   <on_poweroff>destroy</on_poweroff>
>   <on_reboot>restart</on_reboot>
>   <on_crash>restart</on_crash>
>   <devices>
>     <emulator>/vmwrapper</emulator>
>     <disk type='file' device='disk'>
>       <driver name='qemu' type='qcow2'/>
>       <source file='/var/lib/virtlet/volumes/virtlet_root_b28bebf5-38ce-5f2c-686b-2bd4da91f6f9'/>
>       <backingStore type='file' index='1'>
>         <format type='qcow2'/>
>         <source file='/var/lib/virtlet/images/data/e137062a4dfbb4c225971b67781bc52183d14517170e16a3841d16f962ae7470'/>
>         <backingStore/>
>       </backingStore>
>       <target dev='sda' bus='scsi'/>
>       <alias name='scsi0-0-0-0'/>
>       <address type='drive' controller='0' bus='0' target='0' unit='0'/>
>     </disk>
>     <disk type='file' device='cdrom'>
>       <driver name='qemu' type='raw'/>
>       <source file='/var/lib/virtlet/config/config-b28bebf5-38ce-5f2c-686b-2bd4da91f6f9.iso'/>
>       <backingStore/>
>       <target dev='sdb' bus='scsi'/>
>       <readonly/>
>       <alias name='scsi0-0-0-1'/>
>       <address type='drive' controller='0' bus='0' target='0' unit='1'/>
>     </disk>
>     <controller type='scsi' index='0' model='virtio-scsi'>
>       <alias name='scsi0'/>
>       <address type='pci' domain='0x0000' bus='0x00' slot='0x03' function='0x0'/>
>     </controller>
>     <controller type='usb' index='0' model='piix3-uhci'>
>       <alias name='usb'/>
>       <address type='pci' domain='0x0000' bus='0x00' slot='0x01' function='0x2'/>
>     </controller>
>     <controller type='pci' index='0' model='pci-root'>
>       <alias name='pci.0'/>
>     </controller>
>     <serial type='unix'>
>       <source mode='connect' path='/var/lib/libvirt/streamer.sock'>
>         <reconnect enabled='yes' timeout='1'/>
>       </source>
>       <target type='isa-serial' port='0'>
>         <model name='isa-serial'/>
>       </target>
>       <alias name='serial0'/>
>     </serial>
>     <console type='unix'>
>       <source mode='connect' path='/var/lib/libvirt/streamer.sock'>
>         <reconnect enabled='yes' timeout='1'/>
>       </source>
>       <target type='serial' port='0'/>
>       <alias name='serial0'/>
>     </console>
>     <input type='tablet' bus='usb'>
>       <alias name='input0'/>
>       <address type='usb' bus='0' port='1'/>
>     </input>
>     <input type='mouse' bus='ps2'>
>       <alias name='input1'/>
>     </input>
>     <input type='keyboard' bus='ps2'>
>       <alias name='input2'/>
>     </input>
>     <graphics type='vnc' port='5900' autoport='yes' listen='127.0.0.1'>
>       <listen type='address' address='127.0.0.1'/>
>     </graphics>
>     <video>
>       <model type='cirrus' vram='16384' heads='1' primary='yes'/>
>       <alias name='video0'/>
>       <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x0'/>
>     </video>
>     <memballoon model='virtio'>
>       <alias name='balloon0'/>
>       <address type='pci' domain='0x0000' bus='0x00' slot='0x04' function='0x0'/>
>     </memballoon>
>   </devices>
>   <seclabel type='dynamic' model='dac' relabel='yes'>
>     <label>+1000:+101</label>
>     <imagelabel>+1000:+101</imagelabel>
>   </seclabel>
>   <qemu:commandline>
>     <qemu:env name='VIRTLET_EMULATOR' value='/usr/bin/qemu-system-x86_64'/>
>     <qemu:env name='VIRTLET_NET_KEY' value='f2a1a1a5-e070-41ab-be81-7edac5e416fa'/>
>     <qemu:env name='VIRTLET_CONTAINER_ID' value='b28bebf5-38ce-5f2c-686b-2bd4da91f6f9'/>
>     <qemu:env name='VIRTLET_CONTAINER_LOG_PATH' value='/var/log/pods/system_default_vmdefault_f2a1a1a5-e070-41ab-be81-7edac5e416fa/vm/0.log'/>
>     <qemu:env name='CGROUP_PARENT' value='/kubepods/podf2a1a1a5-e070-41ab-be81-7edac5e416fa/b28bebf5-38ce-5f2c-686b-2bd4da91f6f9'/>
>   </qemu:commandline>
> </domain>
> 
> virsh # 
> 
> 
> ======= memory attach hotplug in ==========
> root@ip-172-31-10-115:/# virsh dominfo 1
> Id:             1
> Name:           virtlet-09e7cf24-8210-vm
> UUID:           09e7cf24-8210-57d2-45e9-ca210f20718a
> OS Type:        hvm
> State:          running
> CPU(s):         5
> CPU time:       68.2s
> Max memory:     331776 KiB
> Used memory:    331776 KiB
> Persistent:     yes
> Autostart:      disable
> Managed save:   no
> Security model: none
> Security DOI:   0
> 
> root@ip-172-31-10-115:/# cat mem.xml
> <memory model='dimm'>
>   <target>
>     <size unit='MiB'>128</size>
>     <node>0</node>
>   </target>
> </memory>
> root@ip-172-31-10-115:/# 
>

### CODE REVIEW

File: images/Dockerfile.arktosvmruntime-base [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+MAINTAINER Yunwen Bai <yunwen.bai@futurewei.com>
+
+# BUMP 23.11.2018
+
+ENV DEBIAN_FRONTEND noninteractive
+
+RUN echo deb-src http://archive.ubuntu.com/ubuntu/ bionic main universe restricted >>/etc/apt/sources.list && \
+    echo deb-src http://archive.ubuntu.com/ubuntu/ bionic-updates main universe restricted >>/etc/apt/sources.list
+
+RUN apt-get -y update && apt-get -y upgrade && \
```
Comment (XiaoningDing) [inline images/Dockerfile.arktosvmruntime-base:11]: It's better to combine these RUNs to reduce the layers of the image.

File: pkg/libvirttools/virtualization.go [hunk 2]
```diff
@@ -227,6 +238,21 @@ func (ds *domainSettings) createDomain(config *types.VMConfig) *libvirtxml.Domai
 	return domain
 }
 
+// Helper functions
+// TODO: set max memory and CPU to host allocatable, it is controlled by the CG anyways
```
Comment (XiaoningDing) [inline pkg/libvirttools/virtualization.go:242]: Can you please file an issue so we won't forget this change.

File: images/Dockerfile.arktosvmruntime-base [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+# build libvirt 6.5.0
+RUN apt-get install -y build-essential libsasl2-dev libdevmapper-dev libgnutls28-dev \
+                       libxml2-dev e2fslibs-dev libcap-ng-dev libnl-3-dev \
+                       libselinux1-dev libssh2-1-dev libsystemd-dev \
+                       parted python2.7-dev libreadline-dev libxslt1-dev \
+                       libtirpc-dev libnl-route-3-dev libxml2-utils xsltproc \
+                       python-docutils docbook-xsl uuid-dev
+
+RUN curl -O https://libvirt.org/sources/libvirt-6.5.0.tar.xz && \
```
Comment (XiaoningDing) [inline images/Dockerfile.arktosvmruntime-base:26]: Consider keeping a local copy of the tarball in repo to improve the reliability of the build process?

File: deploy/SetupVmruntime.sh [hunk 4]
```diff
@@ -151,7 +151,6 @@ startRuntime() {
                 DOCKER_RUN_CMD="docker run --security-opt apparmor=${APPARMOR_PROFILE_LIBVIRTD}"
         fi
 	${DOCKER_RUN_CMD} --net=host --privileged --pid=host --uts=host --ipc=host --user=root \
-	--security-opt apparmor=${APPARMOR_PROFILE_LIBVIRTD} \
```
Comment (XiaoningDing) [inline deploy/SetupVmruntime.sh:154]: Why is this removed, is this related to the new version?

Comment (yb01) [reply deploy/SetupVmruntime.sh:154]: this was actually a bug. the DOCKER_RUN_CMD has the proper setting for the --security-opt already, so this line is a dup-- and actually affect the deployment.

File: images/Dockerfile.arktosvmruntime-base (hunk 3 above)
Comment (yb01) [reply images/Dockerfile.arktosvmruntime-base:26]: this is ok here. note that we will push the base image so it won't need to be build so the build process will only build the most top layer of the image. --- this is the current virtlet build process procedure as well.

File: pkg/libvirttools/virtualization.go (hunk 2 above)
Comment (yb01) [reply pkg/libvirttools/virtualization.go:242]: will do. thanks.

File: images/Dockerfile.arktosvmruntime-base (hunk 1 above)
Comment (yb01) [reply images/Dockerfile.arktosvmruntime-base:11]: yes, will do. thanks.

File: pkg/libvirttools/virtualization.go (hunk 2 above)
Comment (yb01) [reply pkg/libvirttools/virtualization.go:242]: issue 44 is created to track this work.

File: images/Dockerfile.arktosvmruntime-base (hunk 1 above)
Comment (yb01) [reply images/Dockerfile.arktosvmruntime-base:11]: fixed by re-org the docker commands and reduced some.
